src/main/puzzle/scenario/scenario-history.gd

```diff
@@ -0,0 +1,116 @@
+class_name ScenarioHistory
+"""
+Stores the best results the player has achieved for different scenarios. This is currently used for calculating high
+scores, but could eventually be used for displaying statistics or calculating rewards.
+"""
+
+# how many daily and all-time records we can store before we start deleting old ones
+var max_size := 3
+
+# key: scenario name
+# value: array of RankResults for the specified scenario
+var rank_results := {}
+
+func scenario_names() -> Array:
+	return rank_results.keys()
+
+
+"""
+Resets the scenario history to its default empty state.
+"""
+func reset() -> void:
+	rank_results.clear()
+
+
+"""
+Returns a player's performances for a specific scenario.
+"""
+func results(scenario: String) -> Array:
+	return rank_results.get(scenario, [])
+
+
+"""
+Returns a player's best performances for a specific scenario.
+
+Parameters:
+	'scenario': The name of the scenario to evaluate
+	
+	'daily': If true, only performances with today's date are included
+	
+	'property': (Optional) The property evaluated to determine which of the player's RankResults was the best, such
+		as 'score' or 'seconds'. Defaults to 'score'.
+"""
+func best_results(scenario: String, daily: bool) -> Array:
+	if not rank_results.has(scenario):
+		return []
+	
+	var results: Array = rank_results[scenario].duplicate()
+	if daily:
+		# only include items with today's date
+		var now := OS.get_datetime()
+		var daily_results := []
+		for result_obj in results:
+			var result: RankResult = result_obj
+			if result.timestamp["year"] == now["year"] \
+					and result.timestamp["month"] == now["month"] \
+					and result.timestamp["day"] == now["day"]:
+				daily_results.append(result)
+		results = daily_results
+	
+	if ScenarioSettings.compare_seconds(scenario):
+		results.sort_custom(self, "_compare_by_seconds")
+	else:
+		results.sort_custom(self, "_compare_by_score")
+	return results.slice(0, max_size - 1)
+
+
+func prev_result(scenario: String) -> RankResult:
+	if not rank_results.has(scenario) or rank_results[scenario].empty():
+		return null
+	return rank_results[scenario][0]
```

Comment:
This method assumes rank_results is ordered by date, but prune() shuffles the array.

Reply:
Surprisingly, prune() was preserving the array's order. Whatever algorithm gdscript's dictionaries use when adding new keys preserves the order in which keys were inserted.

I still changed the algorithm, just in case this behavior changes in the future. It would result in some very confusing and inconsistent bugs.